test/unit-luatest/util_test.lua

```diff
@@ -209,3 +211,44 @@ test_group.test_index = function(g)
     test_index_templ(g, 'memtx')
     test_index_templ(g, 'vinyl')
 end
+
+test_group.test_future_wait_exception = function(g)
+    -- Before 2.10.3 executing future_wait in trigger cause silent hang.
+    t.run_only_if(vutil.version_is_at_least(2, 10, 3, nil, 0, 0))
+    local conn = net.connect(g.server.net_box_uri, {wait_connected = false})
+    t.assert_equals(conn.state, 'initial')
+
+    -- Test that no exceptions're thrown from future_wait
+    local _, err
+    conn:on_connect(function()
+        local f = conn:eval('return true', {}, {is_async = true})
+        _, err = vutil.future_wait(f)
+        require('log').info('exit from on_connect')
+    end)
+
+    conn:wait_connected()
+    t.assert_str_contains(err.message, 'Synchronous requests are not allowed')
+end
+
+test_group.test_future_wait_timeout = function(g)
+    g.server:exec(function()
+        rawset(_G, 'fiber_sleep', function(time)
+            require('fiber').sleep(time)
+            return true
```

Comment:
In the test case below you send an infinite timeout. Even if you later call `future:discard()`, this fiber will keep hanging in `sleep` until restart. I suggest to add a global flag here like: `while _G.do_sleep do fiber.sleep(0.001) end`. Then in the end you do `_G.do_sleep = false` to unblock the fiber and properly wait for the future to end.

The reason why we shouldn't leave such request-fibers is that they might affect global limits like `box.cfg.net_msg_max`.